domain/consensus/utils/consensushashing/calculate_signature_hash.go

```diff
@@ -0,0 +1,196 @@
+package consensushashing
+
+import (
+	"github.com/kaspanet/kaspad/domain/consensus/model/externalapi"
+	"github.com/kaspanet/kaspad/domain/consensus/utils/constants"
+	"github.com/kaspanet/kaspad/domain/consensus/utils/hashes"
+	"github.com/kaspanet/kaspad/domain/consensus/utils/serialization"
+	"github.com/kaspanet/kaspad/domain/consensus/utils/subnetworks"
+	"github.com/pkg/errors"
+)
+
+// SigHashType represents hash type bits at the end of a signature.
+type SigHashType uint32
+
+// Hash type bits from the end of a signature.
+const (
+	SigHashAll          SigHashType = 0x1
+	SigHashNone         SigHashType = 0x2
+	SigHashSingle       SigHashType = 0x3
+	SigHashAnyOneCanPay SigHashType = 0x80
+
+	// SigHashMask defines the number of bits of the hash type which is used
+	// to identify which outputs are signed.
+	SigHashMask = 0x1f
+)
+
+func (sht SigHashType) isSigHashAll() bool {
+	return sht&SigHashMask == SigHashAll
+}
+func (sht SigHashType) isSigHashNone() bool {
+	return sht&SigHashMask == SigHashNone
+}
+func (sht SigHashType) isSigHashSingle() bool {
+	return sht&SigHashMask == SigHashSingle
+}
+func (sht SigHashType) isSigHashAnyOneCanPay() bool {
+	return sht&SigHashAnyOneCanPay == SigHashAnyOneCanPay
```

Comment:
I'm not sure this is correct, I think the correct line is `sht&SigHashAnyOneCanPay != 0`
(This seems to be how bitcoin core does it: https://github.com/bitcoin/bitcoin/blob/8e6532053f580003869346510fc1dc20c46c8067/src/script/interpreter.cpp#L1587 and how we did it before: https://github.com/kaspanet/kaspad/blob/master/domain/consensus/utils/txscript/script.go#L358)

Reply:
Both are correct and widely used.
I prefer the more explicit form of `field&flag == flag`.

Reply:
Hmm you're right haven't thought it through